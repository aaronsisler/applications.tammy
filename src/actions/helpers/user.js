export const setUser = (user) => ({
    type: 'SET_USER',
    user,
});

export const editUser = (updates) => ({
    type: 'EDIT_USER',
    updates,
});

export const clearUser = () => ({
    type: 'CLEAR_USER'
});
