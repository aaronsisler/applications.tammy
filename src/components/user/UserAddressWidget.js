import React from 'react';
import PropTypes from 'prop-types';
import InputTools from 'User/tools/inputs';

export default class UserAddressWidget extends React.Component {
    inputTools = new InputTools();

    constructor(props) {
        super(props);
        this.state = {
            addressLine1: props.user.addressLine1 ? props.user.addressLine1 : '',
            addressLine2: props.user.addressLine2 ? props.user.addressLine2 : '',
            city: props.user.city ? props.user.city : '',
            state: props.user.state ? props.user.state : '',
            postalCode: props.user.postalCode ? props.user.postalCode : '',
        };
    }

    componentDidUpdate(prevProps) {
        const { user: oldUser } = prevProps;
        const { user: newUser } = this.props;

        if (oldUser.userId !== newUser.userId) {
            return this.setState(() => ({
                addressLine1: newUser.addressLine1 ? newUser.addressLine1 : '',
                addressLine2: newUser.addressLine2 ? newUser.addressLine2 : '',
                city: newUser.city ? newUser.city : '',
                state: newUser.state ? newUser.state : '',
                postalCode: newUser.postalCode ? newUser.postalCode : '',
            }));
        }
    }

    handleInputChange = (e) => {
        const inputName = e.target.name;
        const inputValue = e.target.value;

        switch (inputName) {
            case 'postalCode':
                this.handlePostalCodeValidation(inputValue);
                return;
            case 'state':
                this.handleStateAbbreviationValidation(inputValue);
                return;
            default:
                this.handleAddInput(inputName, inputValue);
                return;
        }
    }

    handlePostalCodeValidation = (inputValue) => {
        if (!inputValue || inputValue.match(/^\d{1,5}?$/)) {
            this.setState(() => ({ postalCode: inputValue }));
        }
    }

    handleStateAbbreviationValidation = (inputValue) => {
        if (!inputValue || inputValue.match(/^[A-Za-z]{1,2}$/)) {
            this.setState(() => ({ state: inputValue.toUpperCase() }));
        }
    }

    handleAddInput = (inputName, inputValue) => (
        this.setState(() => ({
            [inputName]: inputValue
        }))
    )

    handleSubmit = () => {
        const submitObject = Object.assign({}, this.state);
        this.props.onSubmit(submitObject);
    }

    render() {
        return (
            <div className="user_address_widget">
                <div className="user_address_content">
                    <div className="user_address_widget__title">
                        Address
                    </div>
                    <div className="user_address_widget__inputs">
                        <input
                            readOnly={this.props.isReadOnly}
                            type="text"
                            name="addressLine1"
                            placeholder="Address Line 1"
                            value={this.state.addressLine1}
                            onChange={this.handleInputChange}
                            onBlur={this.inputTools.handleRequiredValidation}
                        />
                        <input
                            readOnly={this.props.isReadOnly}
                            type="text"
                            name="addressLine2"
                            placeholder="Address Line 2"
                            value={this.state.addressLine2}
                            onChange={this.handleInputChange}
                        />
                        <div className="user_address_widget__inputs_location">
                            <input
                                readOnly={this.props.isReadOnly}
                                type="text"
                                name="city"
                                placeholder="City"
                                value={this.state.city}
                                onChange={this.handleInputChange}
                                onBlur={this.inputTools.handleRequiredValidation}
                            />
                            <input
                                readOnly={this.props.isReadOnly}
                                type="text"
                                name="state"
                                placeholder="State"
                                value={this.state.state}
                                onChange={this.handleInputChange}
                                onBlur={this.inputTools.handleRequiredValidation}
                            />
                            <input
                                readOnly={this.props.isReadOnly}
                                type="text"
                                name="postalCode"
                                placeholder="Postal Code"
                                value={this.state.postalCode}
                                onChange={this.handleInputChange}
                                onBlur={this.inputTools.handleRequiredValidation}
                            />
                        </div>
                    </div>
                </div>
                {!this.props.isReadOnly &&
                    <div className="user_address_widget__button">
                        <button
                            disabled={!this.state.addressLine1 || !this.state.city || !this.state.state || !this.state.postalCode}
                            onClick={this.handleSubmit}
                        >
                            Update Address
                       </button>
                    </div>
                }
            </div>
        )
    }
}

UserAddressWidget.propTypes = {
    isReadOnly: PropTypes.bool,
    user: PropTypes.object.isRequired,
    onSubmit: PropTypes.func,
};
