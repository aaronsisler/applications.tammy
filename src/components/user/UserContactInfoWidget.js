import React from 'react';
import PropTypes from 'prop-types';
import { requiredInputFieldClassName } from '../../tools/constants';
import { isEmailValid } from '../../tools/email';

export default class UserNameWidget extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isEmailValid: props.user.email ? true : false,
            email: props.user.email ? props.user.email : '',
            phoneAreaCode: props.user.phoneAreaCode ? props.user.phoneAreaCode : '',
            phonePrefix: props.user.phonePrefix ? props.user.phonePrefix : '',
            phoneLineNumber: props.user.phoneLineNumber ? props.user.phoneLineNumber : '',
            phoneExt: props.user.phoneExt ? props.user.phoneExt : '',
        };
    }

    handleEmailInput = (e) => {
        const inputValue = e.target.value;

        this.setState({ email: inputValue });
    }

    handleEmailValidation = (e) => {
        const inputName = e.target.name;
        const inputValue = e.target.value;

        if (isEmailValid(inputValue)) {
            document.getElementsByName(inputName)[0].classList.remove(requiredInputFieldClassName);
            return this.setState({ isEmailValid: true });
        }
        else {
            document.getElementsByName(inputName)[0].classList.add(requiredInputFieldClassName);
            return this.setState({ isEmailValid: false });
        }
    }

    handlePhoneNumberValidation = (e) => {
        const inputName = e.target.name;
        const inputValue = e.target.value;

        if (!inputValue) {
            return this.setState({ [inputName]: inputValue });
        }

        switch (inputName) {
            case 'phoneAreaCode':
            case 'phonePrefix':
                if (inputValue.match(/^\d{1,3}?$/)) {
                    this.setState({ [inputName]: inputValue });
                }
                return;
            case 'phoneLineNumber':
                if (inputValue.match(/^\d{1,4}?$/)) {
                    this.setState({ [inputName]: inputValue });
                }
                return;
            default:
                if (inputValue.match(/^\d{1,5}?$/)) {
                    this.setState({ [inputName]: inputValue });
                }
                return;
        }
    }

    handleSubmit = () => {
        const submitObject = Object.assign({}, this.state);
        this.props.onSubmit(submitObject);
    }

    render() {
        return (
            <div id="user_contact_info_widget">
                <div className="user_email_content" >
                    <div className="user_contact_info_content_title">
                        Email
                    </div>
                    <div className="user_email_input">
                        <input
                            readOnly={this.props.isReadOnly}
                            type="text"
                            name="email"
                            placeholder="Email"
                            className="text_input"
                            value={this.state.email}
                            onChange={this.handleEmailInput}
                            onBlur={this.handleEmailValidation}
                        />
                    </div>
                </div>
                <div className="user_phone_number_content">
                    <div className="user_contact_info_content_title">
                        Phone Number
                    </div>
                    <div className="user_phone_number_inputs">
                        <div className="user_phone_number_input">
                            <input
                                readOnly={this.props.isReadOnly}
                                type="text"
                                name="phoneAreaCode"
                                placeholder="(919)"
                                className="text_input"
                                value={this.state.phoneAreaCode}
                                onChange={this.handlePhoneNumberValidation}
                            />
                        </div>
                        <div className="user_phone_number_input">
                            <input
                                readOnly={this.props.isReadOnly}
                                type="text"
                                name="phonePrefix"
                                placeholder="123"
                                className="text_input"
                                value={this.state.phonePrefix}
                                onChange={this.handlePhoneNumberValidation}
                            />
                        </div>
                        <div className="user_phone_number_input">
                            <input
                                readOnly={this.props.isReadOnly}
                                type="text"
                                name="phoneLineNumber"
                                placeholder="4567"
                                className="text_input"
                                value={this.state.phoneLineNumber}
                                onChange={this.handlePhoneNumberValidation}
                            />
                        </div>
                        <div className="user_phone_number_input">
                            <input
                                readOnly={this.props.isReadOnly}
                                type="text"
                                name="phoneExt"
                                placeholder="Ext."
                                className="text_input"
                                value={this.state.phoneExt}
                                onChange={this.handlePhoneNumberValidation}
                            />
                        </div>
                    </div>
                </div>
                <div className="user_contact_info_widget_button">
                    {!this.props.isReadOnly &&
                        <button
                            disabled={!this.state.isEmailValid}
                            onClick={this.handleSubmit}
                            className="button"
                        >
                            Update Contact Info
                    </button>
                    }
                </div>
            </div>
        )
    }
}

UserNameWidget.propTypes = {
    isReadOnly: PropTypes.bool,
    user: PropTypes.object.isRequired,
    onSubmit: PropTypes.func.isRequired,
};
