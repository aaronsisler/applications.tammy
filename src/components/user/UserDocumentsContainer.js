import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import DocumentUploadWidget from '../shared/document/DocumentUploadWidget';
import DocumentList from '../shared/document/DocumentList';
import LoadingPage from '../core/LoadingPage';

export class UserDocumentsContainer extends React.Component {
    constructor(props) {
        super(props);
    }

    componentDidMount() {
    }

    render() {
        return (
            <div>
                {!this.props.documents && <LoadingPage />}
                {this.props.documents &&
                    <div id="user_documents_container">
                        <DocumentUploadWidget />
                        <DocumentList
                            documents={this.props.documents}
                        />
                    </div>
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    documents: state.userDocuments,
})

export default connect(mapStateToProps)(UserDocumentsContainer);

UserDocumentsContainer.propTypes = {
    documents: PropTypes.array.isRequired,
};
