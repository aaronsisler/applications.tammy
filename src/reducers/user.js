const userReducerDefaultState = null;

const userReducer = (state = userReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_USER':
            return action.user;
        case 'CLEAR_USER':
            return null;
        case 'EDIT_USER':
            return {
                ...state,
                ...action.updates
            }
        default:
            return state;
    }
};

export default userReducer;

