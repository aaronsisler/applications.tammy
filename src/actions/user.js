import database from 'Firebase/firebase';
import { setUser, editUser, clearUser } from './helpers/user';

export const startSetUser = () => (dispatch, getState) => {
    const { uid: userId } = getState().auth;

    return database.ref(`users/${userId}`).once('value').then((snapshot) => {
        dispatch(setUser(
            {
                userId: snapshot.key,
                ...snapshot.val()
            }));
    })
};

export const startEditUser = (updates) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;

    return database.ref(`users/${userId}`).update(updates).then(() => {
        dispatch(editUser(updates));
    });
};

export const startClearUser = () => (dispatch) => dispatch(clearUser());
