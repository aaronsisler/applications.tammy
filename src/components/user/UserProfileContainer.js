import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import UserNameWidget from './UserNameWidget';
import UserAddressWidget from './UserAddressWidget';
import UserContactInfoWidget from './UserContactInfoWidget';
import { startEditUser } from '../../actions/user';
import LoadingPage from '../core/LoadingPage';

export class UserProfileContainer extends React.Component {
    constructor(props) {
        super(props);
    }

    onSubmit = (user) => {
        this.props.startEditUser(user);
    }
    render() {
        return (
            <div>
                {!this.props.user && <LoadingPage />}
                {this.props.user &&
                    <div id="user_profile_container">
                        <div className="user_widgets_column">
                            <div className="user_widget">
                                <UserNameWidget
                                    user={this.props.user}
                                    onSubmit={this.onSubmit}
                                />
                            </div>
                            <div className="user_widget">
                                <UserAddressWidget
                                    user={this.props.user}
                                    onSubmit={this.onSubmit}
                                />
                            </div>
                        </div>
                        <div className="user_widgets_column">
                            <div className="user_widget">
                                <UserContactInfoWidget
                                    user={this.props.user}
                                    onSubmit={this.onSubmit}
                                />
                            </div>
                        </div>
                    </div>
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    user: state.user,
})

const mapDispatchToProps = (dispatch) => ({
    startEditUser: (user) => dispatch(startEditUser(user)),
})

export default connect(mapStateToProps, mapDispatchToProps)(UserProfileContainer);

UserProfileContainer.propTypes = {
    user: PropTypes.object.isRequired,
    startEditUser: PropTypes.func.isRequired,
};
