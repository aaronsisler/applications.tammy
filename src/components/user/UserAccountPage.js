import React from 'react';
import UserProfileContainer from './UserProfileContainer';

export default class UserAccountPage extends React.Component {
    constructor(props) {
        super(props);
    }

    componentDidMount() {
    }

    render() {
        return (
            <div id="user_account_page">
                <UserProfileContainer />
            </div>
        );
    }
}
