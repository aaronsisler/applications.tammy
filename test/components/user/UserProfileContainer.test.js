
import React from 'react';
import { shallow } from 'enzyme';
import { UserProfileContainer } from 'User/UserProfileContainer';
import user from '../../fixtures/user';

describe('UserProfileContainer', () => {
    const startEditUser = jest.fn();
    let wrapper;

    const buildWrapper = (isReadOnly, inputUser) => {
        wrapper = shallow(
            <UserProfileContainer
                isReadOnly={isReadOnly}
                startEditUser={startEditUser}
                user={inputUser}
            />);
    }

    describe('when user is defined', () => {
        beforeEach(() => {
            buildWrapper(false, user);
        });

        it('should render UserProfileContainer correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });

        describe('startEditUser() prop', () => {
            it('should be called for UserNameWidget`s onSubmit prop', () => {
                wrapper.find('#UserNameWidget').prop('onSubmit')(user);

                expect(startEditUser).toHaveBeenLastCalledWith(user);
            });

            it('should be passed as UserContactInfoWidget`s onSubmit prop', () => {
                wrapper.find('#UserContactInfoWidget').prop('onSubmit')(user);

                expect(startEditUser).toHaveBeenLastCalledWith(user);
            });

            it('should be passed as UserAddressWidget`s onSubmit prop', () => {
                wrapper.find('#UserAddressWidget').prop('onSubmit')(user);

                expect(startEditUser).toHaveBeenLastCalledWith(user);
            });
        });
    });

    describe('when user is undefined', () => {
        beforeEach(() => {
            buildWrapper(false, undefined);
        });

        it('should render UserProfileContainer correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
