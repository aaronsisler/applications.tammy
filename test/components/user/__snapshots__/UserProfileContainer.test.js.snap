// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileContainer when user is defined should render UserProfileContainer correctly 1`] = `
<div>
  <div
    id="user_profile_container"
  >
    <div
      className="user_widget"
    >
      <UserNameWidget
        id="UserNameWidget"
        isReadOnly={false}
        onSubmit={[Function]}
        user={
          Object {
            "addressLine1": "123 Evergreen Terrace",
            "addressLine2": "Apt. 456",
            "city": "Durham",
            "displayName": "Johnny Appleseed",
            "email": "johnny.appleseed@gmail.com",
            "firstName": "Johnny",
            "lastName": "Appleseed",
            "middleName": "Middleton",
            "phoneAreaCode": "919",
            "phoneExt": "12345",
            "phoneLineNumber": "4567",
            "phonePrefix": "123",
            "postalCode": "27540",
            "state": "NC",
          }
        }
      />
    </div>
    <div
      className="user_widget"
    >
      <UserNameWidget
        id="UserContactInfoWidget"
        isReadOnly={false}
        onSubmit={[Function]}
        user={
          Object {
            "addressLine1": "123 Evergreen Terrace",
            "addressLine2": "Apt. 456",
            "city": "Durham",
            "displayName": "Johnny Appleseed",
            "email": "johnny.appleseed@gmail.com",
            "firstName": "Johnny",
            "lastName": "Appleseed",
            "middleName": "Middleton",
            "phoneAreaCode": "919",
            "phoneExt": "12345",
            "phoneLineNumber": "4567",
            "phonePrefix": "123",
            "postalCode": "27540",
            "state": "NC",
          }
        }
      />
    </div>
    <div
      className="user_widget"
    >
      <UserAddressWidget
        id="UserAddressWidget"
        isReadOnly={false}
        onSubmit={[Function]}
        user={
          Object {
            "addressLine1": "123 Evergreen Terrace",
            "addressLine2": "Apt. 456",
            "city": "Durham",
            "displayName": "Johnny Appleseed",
            "email": "johnny.appleseed@gmail.com",
            "firstName": "Johnny",
            "lastName": "Appleseed",
            "middleName": "Middleton",
            "phoneAreaCode": "919",
            "phoneExt": "12345",
            "phoneLineNumber": "4567",
            "phonePrefix": "123",
            "postalCode": "27540",
            "state": "NC",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`UserProfileContainer when user is undefined should render UserProfileContainer correctly 1`] = `
<div>
  <LoadingPage />
</div>
`;
